File: ruby-agent/inators/filerw.rb
module Bu
	class FileReadWriteInator
		def initialize
			@on_read = lambda { |file, length, offset, data| puts "file_rw@on_read file => #{file} length => #{length} offset => #{offset} data => #{data}" }
			@on_write = lambda { |file, data, offset, length| puts "file_rw@on_write file => #{file} data => #{data} offset => #{offset} data => #{data}" }
			@on_error = lambda { |file, mode, error| puts "file_rw@on_error file => #{file} mode => #{mode} error => #{error}" }
		end

		def on(event, &callback)
			case event
			when :read
				@on_read = callback
			when :write
				@on_write = callback
			when :error
				@on_error = callback
			end
		end

		def read(file, length, offset)
			begin
				data = File.binread(file, length, offset)
				@on_read.call(file, length, offset, data)
			rescue StandardError => msg
				@on_error.call(file, :read, msg)
			end
		end

		def write(file, data, offset)
			begin
				length = File.binwrite(file, data, offset)
				@on_write.call(file, data, offset, length)
			rescue StandardError => msg
				@on_error.call(file, :write, msg)
			end
		end
	end
end

File: go-agent/inators/filerw/filerw.go
package filerw

import(
	_"os"
	"fmt"
)

type FileReadWriteInator interface {
	Read(file string, length int, offset int)
	Write(file string, data []byte, offset int)
	OnRead(callback func (file string, length int, offset int, data []byte))
	OnWrite(callback func (file string, data []byte, offset int, length int))
	OnError(callback func (file string, mode string, error string))
}

type fileReadWriteInator struct {
	onRead		func (file string, length int, offset int, data []byte)
	onWrite		func (file string, data []byte, offset int, length int)
	onError		func (file string, mode string, error string)
}

func NewFileReadWriteInator() *fileReadWriteInator {
	frw := &fileReadWriteInator {
		onRead 	: func (file string, length int, offset int, data []byte) { fmt.Println("filerw@read ") },
		onWrite : func (file string, data []byte, offset int, length int) { fmt.Println("filerw@write ") },
		onError	: func (file string, mode string, error string) { fmt.Println("filerw@error ") },
	}
	return frw
}

func (frw *fileReadWriteInator) Read(file string, length int, offset int) {
	data := []byte("ONE OK ROCK") //test data
	frw.onRead(file, length, offset, data)
}

func (frw *fileReadWriteInator) Write(file string, data []byte, offset int) {
	length := 10969	//test length
	frw.onWrite(file, data, offset, length)
}

func (frw *fileReadWriteInator) OnRead(callback func (file string, length int, offset int, data []byte)) {
	frw.onRead = callback 
}

func (frw *fileReadWriteInator) OnWrite(callback func (file string, data []byte, offset int, length int)) {
	frw.onWrite = callback 
}

func (frw *fileReadWriteInator) OnError(callback func (file string, mode string, error string)) {
	frw.onError = callback 
}

